### DISCUSSION TITLE
Error rendering data template - Solved!

### DISCUSSION THREAD

**Discussion by michelcmorel** [HASwitchPlate/HASPone#89] [General]
Update: this error is solved just so you know at the bottom of this page are my findings.

Hi there,
After I run the core functionality blueprint the first time, save and run action. Each time i get:
Error rendering data template: UndefinedError: None has no element 0
In the home assistant logs I get the following:
2021-04-28 00:48:41 ERROR (MainThread) [homeassistant.components.automation.hasp_core_functionality] HASP Core functionality: Choose at step 1: choice 1: Error executing script. Error for call_service at pos 8: Error rendering data template: UndefinedError: None has no element 0
2021-04-28 00:48:41 ERROR (MainThread) [homeassistant.components.automation.hasp_core_functionality] HASP Core functionality: Error executing script. Error for choose at pos 1: Error rendering data template: UndefinedError: None has no element 0
2021-04-28 00:48:41 ERROR (MainThread) [homeassistant.components.automation.hasp_core_functionality] Error while executing automation automation.hasp_core_functionality: Error rendering data template: UndefinedError: None has no element 0
Would anyone know if this is an issue with my setup or any pointers would be appreciated please?
Thanks!

28.4.21 update
I noticed also that when I click on run action, I did not get any notification of the yaml code. But MQTT has discovered the device. The arrow buttons to change the page are not visible meaning they do not appear (I guess since I can't see those initial page setup on the screen that makes sense), and the nextion page has the wifi connected & MQTT connected text.
I tried to roll back and ran a snapshot and did all the steps again to get to the same bottle neck issue.
![Screenshot 2021-04-28 at 08 28 19](https://user-images.githubusercontent.com/4426933/116367056-d4e15100-a7fe-11eb-8674-6887f6ba93ac.png)

In the trace timeline it looks like the script did not have a chance to start.
![Screenshot 2021-04-28 at 12 49 41](https://user-images.githubusercontent.com/4426933/116399070-6b723a00-a820-11eb-9818-93b64bb20c92.png)

Update:
I think I have found the source of the error, anyone would know this error code?
![Screenshot 2021-04-28 at 14 18 25](https://user-images.githubusercontent.com/4426933/116410663-e5102500-a82c-11eb-988e-8598cfe81f15.png)

Update:
So I solved this issue and might as well share my findings in case this can help someone else.

To start with I needed to get more debug information so to troubleshoot the issue better. I'm on a mac so I tried to "telnet" in my arduino IP, that did not work. I tried nc (netcat) in terminal that did not work for me even though I could connect successfully I could not get any debug info to appear. I ended up using esphomeflasher to read the debug log information, since my arduino is connected to my mac using usb connection the esphomeflasher will display the debug log info if you click on view logs.
I found that there was some error between my nextion and arduino so I figured out to start again with flashing nextion using an old pc latpop that i used initially to format the micro sd card. So I did that again, flash it. (Just use the red and black wires to flash it, do not connnect the communication wires (yellow, blue) to arduino yet. Then once the flash is done I connected the yellow to D4  & blue wire to D7 on the arduino. Then I rebooted everything, HA, arduino and the nextion screen.
This time in esphomeflasher logs I could see that the communication was established correctly without any errors. I then went on the import core functionality blueprint on haspone repo and then save, ran the blueprint, the logs show that the nextion was receiving the information without any errors, but when looking at my home assistant config logs (/config/home-assistant.log) I was still getting the same error "2021-04-28 20:57:26 ERROR (MainThread) [homeassistant.components.automation.hasp_core_functionality_4] Error while executing automation automation.hasp_core_functionality_4: Error rendering data template: UndefinedError: None has no element 0".
So I went on haspone github page and this time imported the black template (HASP Theme Light on Dark) located at https://github.com/HASwitchPlate/HASPone/tree/main/Home_Assistant/blueprints. then ran the blueprint, but still nothing appeared, at that point I started tapping on the nextion screen and the page just appeared.
And now it appears to be working, I ran a lights on/off and working well :)

**Comment by richlee91** [comment]
I am having a similar problem.  I bought a HASPOne premade unit.  I got it connected and HA setup with the blueprint installed when I runactions I get 
Executed: May 1, 2021, 12:09:51 PM
Error: Error rendering data template: UndefinedError: None has no element 0
In the stack trace.  I am not sure about altering the communications wires as described and I thought the core was supposed to install before the templates.  
![image](https://user-images.githubusercontent.com/24630720/116788322-f0ef2780-aa76-11eb-894c-e91f1da01633.png)

Can anyone please advise?
Thanks

**Reply by michelcmorel** [reply]
> I am having a similar problem. I bought a HASPOne premade unit. I got it connected and HA setup with the blueprint installed when I runactions I get
> Executed: May 1, 2021, 12:09:51 PM
> Error: Error rendering data template: UndefinedError: None has no element 0
> In the stack trace. I am not sure about altering the communications wires as described and I thought the core was supposed to install before the templates.
> ![image](https://user-images.githubusercontent.com/24630720/116788322-f0ef2780-aa76-11eb-894c-e91f1da01633.png)
> 
> Can anyone please advise?
> Thanks

I'm not sure how the pre-made units are built sorry, meaning I don't know if both devices needs flashing or have been flashed already (check out the documentation). The issue that I had was related around the nextion display and the blueprints.
Did you buy the pre-made units on Luma's Tindie page? https://www.tindie.com/products/luma/ha-switchplate-hasp-single-wide-assembled/
I think there is a contact button on that page to reach the maker (Luma's store), and he might be able to advise / help.
I built my own version of HaspOne and followed this documentation is:  https://github.com/HASwitchPlate/HASPone and https://github.com/HASwitchPlate/HASPone/wiki/Configure-your-HASPone-for-Home-Assistant
I had to flash both the nextion display and the arduino d1 mini module (..a few times, perhaps I wasn't doing the flashing correctly as the steps are really important to follow, and then i imported the core functionality blueprint (check the videos on youtube), then (in my case) I had to import "HASP Theme Light on Dark" so to get rid of that error, the link for that blueprint is https://github.com/HASwitchPlate/HASPone/tree/main/Home_Assistant/blueprints#hasp-theme-light-on-dark). If you have a look on youtube there are lots of videos explaining the setup process (search for "ha switch plate" or "HaspOne"). But bare in mind that the process may be slightly different if you are using HaspOne code and from what I can gather is a version up from the original ha switch plate.
Yeah your error seems similar to what I had. Not sure how much tinkering you do but in my case I had to import a few blueprints and flash my nextion and arduino D1 mini a few times before that error disappeared. It took me a few days to work out the few issues that I had but now it's working really well. I've even added buttons on my nextion display to play online radios from my echo dot by tapping on the screen / button.

**Reply by aderusha** [reply]
I've had a few reports of issues that look like this and am trying to [chase some of that down over here](https://github.com/HASwitchPlate/HASPone/discussions/93).
